File: golang/sieve.go
package main

import (
	"math"
)

type Sieve struct {
	BaseBenchmark
	limit    int64
	checksum uint32
}

func (s *Sieve) Prepare() {
	s.limit = s.ConfigVal("limit")
	s.checksum = 0
}

func (s *Sieve) Run(iteration_id int) {
	lim := int(s.limit)
	primes := make([]byte, lim+1)
	for i := 0; i <= lim; i++ {
		primes[i] = 1
	}
	primes[0] = 0
	primes[1] = 0

	sqrtLimit := int(math.Sqrt(float64(lim)))

	for p := 2; p <= sqrtLimit; p++ {
		if primes[p] == 1 {
			for multiple := p * p; multiple <= lim; multiple += p {
				primes[multiple] = 0
			}
		}
	}

	lastPrime := 2
	count := 1

	for n := 3; n <= lim; n += 2 {
		if primes[n] == 1 {
			lastPrime = n
			count++
		}
	}

	s.checksum += uint32(lastPrime + count)
}

func (s *Sieve) Checksum() uint32 {
	return s.checksum
}

File: cpp/src/sieve.cpp
#include "sieve.hpp"
#include <cmath>
#include <vector>

Sieve::Sieve() : limit(config_val("limit")), checksum_val(0) {}

std::string Sieve::name() const { return "Etc::Sieve"; }

void Sieve::run(int iteration_id) {
  (void)iteration_id;
  size_t sz = static_cast<size_t>(limit);
  std::vector<uint8_t> primes(sz + 1, 1);
  primes[0] = 0;
  primes[1] = 0;

  size_t sqrt_limit =
      static_cast<size_t>(std::sqrt(static_cast<double>(limit)));

  for (size_t p = 2; p <= sqrt_limit; ++p) {
    if (primes[p] == 1) {
      for (size_t multiple = p * p; multiple <= sz; multiple += p) {
        primes[multiple] = 0;
      }
    }
  }

  int last_prime = 2;
  int count = 1;

  for (size_t n = 3; n <= sz; n += 2) {
    if (primes[n] == 1) {
      last_prime = static_cast<int>(n);
      count++;
    }
  }

  checksum_val += static_cast<uint32_t>(last_prime + count);
}

uint32_t Sieve::checksum() { return checksum_val; }